packages/engine/lib/nano/src/client.rs

```diff
@@ -113,15 +133,17 @@ impl Client {
     ///
     /// # Errors
     ///
-    /// Creating a `Client` returns an error if
+    /// Creating a `Client` returns [`ErrorKind::ClientCreation`] if
     ///
     /// - a worker could not be created from the provided url.
     pub fn new(url: &str, num_workers: usize) -> Result<Self> {
         let (sender, receiver) = spmc::channel(num_workers);
 
         let workers = (0..num_workers)
             .map(|_| {
-                let mut worker = Worker::new(url, receiver.clone())?;
+                let mut worker = Worker::new(url, receiver.clone())
+                    .wrap_err("Could not create nng worker")
```

Comment:
Similar question to above, if `Worker::new` fails does it return an error that basically says "Couldn't create worker"? Because if so I don't think we're gaining anything by adding this information here.

Reply:
Okay I've looked at the code and I also see that this is actually the one that's in your PR description (and the message seems necessary as without it we would actually be missing a lot of information).

I think we're adding context at the wrong level here though, shouldn't the `Worker::new` impl should be the thing adding the "Couldn't create worker" error message? It doesn't seem like we have any more contextual information at this level that we're actually adding.

Reply:
I agree that adding the context would be better at the `Worker::new` implementation. However, then every function would need to add this function. The easy way out would be [`try` blocks](https://doc.rust-lang.org/beta/unstable-book/language-features/try-blocks.html), which are still unstable. They are implemented but type inference doesn't work right now. AFAIK this is the only bug left for `try` blocks.

Reply:
Generally, it seems that the Rust ecosystem puts the error context **on** the function, not **in** the function, but most probably because of the issue I just mentioned.

Reply:
Yeah that makes sense. I think we could fix this with a macro but I'm not sure it's the right way to go. If we're just generally trying to wrap all of the code _in_ a function with a `try` then I don't think we needed to add context in the first place because that should be part of the backtrace.

e.g. `worker::new` doesn't need every message to say "couldn't create worker" because that's in the backtrace.

Reply:
> I think we could fix this with a macro but I'm not sure it's the right way to go

I also thought about that but I don't like it thus I didn't mention it 😄

Reply:
Perhaps we want to change the `caused by` formatting to include the current level of the stack? Because we basically want this:
```
   2: nano::client::Worker::new
             at ./lib/nano/src/client.rs:60:22
```